src/sundials/sundatanode/sundatanode_inmem.c

```diff
@@ -0,0 +1,475 @@
+
+SUNErrCode SUNDataNode_GetDataNvector_InMem(const SUNDataNode self, N_Vector v,
+                                            sunrealtype* t)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNMemory leaf_data = (SUNMemory)IMPL_MEMBER(self, leaf_data);
+
+  SUNMemoryType leaf_mem_type = leaf_data->type;
+  SUNMemoryType buffer_mem_type = N_VGetDeviceArrayPointer(v) ? SUNMEMTYPE_DEVICE
+                                                              : SUNMEMTYPE_HOST;
+
+  sunindextype buffer_size = 0;
+  SUNCheckCall(N_VBufSize(v, &buffer_size));
+  SUNAssert((buffer_size + sizeof(sunrealtype)) == leaf_data->bytes,
+            SUN_ERR_ARG_INCOMPATIBLE);
+
+  if (leaf_mem_type == buffer_mem_type)
+  {
+    sunrealtype* data_ptr = leaf_data->ptr;
+    *t                    = data_ptr[0];
+    SUNCheckCall(N_VBufUnpack(v, &data_ptr[1]));
+  }
+  else
+  {
+    SUNMemory buffer_data = NULL;
+    SUNCheckCall(SUNMemoryHelper_Alloc(IMPL_MEMBER(self, mem_helper), &buffer_data,
+                                       buffer_size, buffer_mem_type, queue));
+
+    SUNCheckCall(SUNMemoryHelper_Copy(IMPL_MEMBER(self, mem_helper), buffer_data,
+                                      leaf_data, buffer_size, queue));
+
+    sunrealtype* data_ptr = leaf_data->ptr;
+    *t                    = data_ptr[0];
+    data_ptr              = buffer_data->ptr;
+
+    SUNCheckCall(N_VBufUnpack(v, &data_ptr[1]));
+
+    SUNCheckCall(SUNMemoryHelper_Dealloc(IMPL_MEMBER(self, mem_helper),
+                                         buffer_data, queue));
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_SetData_InMem(SUNDataNode self, SUNMemoryType src_mem_type,
+                                     SUNMemoryType node_mem_type, void* data,
+                                     size_t data_stride, size_t data_bytes)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNAssert(BASE_MEMBER(self, dtype) == SUNDATANODE_LEAF, SUN_ERR_ARG_WRONGTYPE);
+
+  SUNMemory data_mem_src = SUNMemoryHelper_Wrap(IMPL_MEMBER(self, mem_helper),
+                                                data, src_mem_type);
+  SUNCheckLastErr();
+
+  SUNMemory data_mem_dst = NULL;
+  SUNCheckCall(SUNMemoryHelper_AllocStrided(IMPL_MEMBER(self, mem_helper),
+                                            &data_mem_dst, data_bytes,
+                                            data_stride, node_mem_type, queue));
+
+  SUNCheckCall(SUNMemoryHelper_Copy(IMPL_MEMBER(self, mem_helper), data_mem_dst,
+                                    data_mem_src, data_bytes, queue));
+
+  SUNMemoryHelper_Dealloc(IMPL_MEMBER(self, mem_helper), data_mem_src, queue);
+
+  IMPL_MEMBER(self, leaf_data) = data_mem_dst;
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_SetDataNvector_InMem(SUNDataNode self, N_Vector v,
+                                            sunrealtype t)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNMemoryType leaf_mem_type = SUNMEMTYPE_HOST;
```

Comment:
Should this be `leaf_data->type`?

Reply:
No, we are storing the leaf data in the host memory.